commands/command_migrate_export.go

```diff
@@ -0,0 +1,203 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/git-lfs/git-lfs/filepathfilter"
+	"github.com/git-lfs/git-lfs/git"
+	"github.com/git-lfs/git-lfs/git/githistory"
+	"github.com/git-lfs/git-lfs/git/odb"
+	"github.com/git-lfs/git-lfs/lfs"
+	"github.com/git-lfs/git-lfs/tasklog"
+	"github.com/git-lfs/git-lfs/tools"
+	"github.com/spf13/cobra"
+)
+
+func migrateExportCommand(cmd *cobra.Command, args []string) {
+	l := tasklog.NewLogger(os.Stderr)
+	defer l.Close()
+
+	db, err := getObjectDatabase()
+	if err != nil {
+		ExitWithError(err)
+	}
+	defer db.Close()
+
+	rewriter := getHistoryRewriter(cmd, db, l)
+
+	filter := rewriter.Filter()
+	if len(filter.Include()) <= 0 {
+		ExitWithError(errors.Errorf("fatal: one or more files must be specified with --include"))
+	}
+
+	tracked := trackedFromExportFilter(filter)
+	gitfilter := lfs.NewGitFilter(cfg)
+
+	var exported []string
+
+	opts := &githistory.RewriteOptions{
+		Verbose:           migrateVerbose,
+		ObjectMapFilePath: objectMapFilePath,
+		BlobFn: func(path string, b *odb.Blob) (*odb.Blob, error) {
+			if filepath.Base(path) == ".gitattributes" {
+				return b, nil
+			}
+
+			ptr, err := lfs.DecodePointer(b.Contents)
+			if err != nil {
+				if errors.IsNotAPointerError(err) {
+					return b, nil
+				}
+				return nil, err
+			}
+
+			downloadPath, err := gitfilter.ObjectPath(ptr.Oid)
+			if err != nil {
+				return nil, err
+			}
+
+			newBlob, err := odb.NewBlobFromFile(downloadPath)
+			if err != nil {
+				return nil, err
+			}
+
+			exported = append(exported, downloadPath)
+			return newBlob, nil
+		},
+
+		TreeCallbackFn: func(path string, t *odb.Tree) (*odb.Tree, error) {
+			if path != "/" {
+				// Ignore non-root trees.
+				return t, nil
+			}
+
+			ours := tracked
+			theirs, err := trackedFromAttrs(db, t)
+			if err != nil {
+				return nil, err
+			}
+
+			// Create a blob of the attributes that are optionally
+			// present in the "t" tree's .gitattributes blob, and
+			// union in the patterns that we've tracked.
+			//
+			// Perform this Union() operation each time we visit a
+			// root tree such that if the underlying .gitattributes
+			// is present and has a diff between commits in the
+			// range of commits to migrate, those changes are
+			// preserved.
+			blob, err := trackedToBlob(db, theirs.Clone().Union(ours))
```

Comment:
It might if there was also no pre-existing `.gitattributes` file, but right now we require a pattern being given to `migrate export` through the `--include` argument, so it shouldn't be an issue.

Reply:
Thanks for the clarification, that sounds good to me.